layer/layer_store.go

```diff
@@ -255,7 +254,7 @@ func (ls *layerStore) registerWithDescriptor(ts io.Reader, parent ChainID, descr
 		pid = p.cacheID
 		// Release parent chain if error
 		defer func() {
-			if err != nil {
+			if err != nil || exist {
```

Comment:
This isn't where the cleanup needs to happen. If the logic is already correct to the clean up the layer there is no need to make a code change.